_pytest/skipping.py

```diff
@@ -107,40 +116,51 @@ def _getglobals(self):
-                    or [self.holder.mark]
-                for _, args, kwargs in marks:
-                    if 'condition' in kwargs:
-                        args = (kwargs['condition'],)
-                    for expr in args:
+        self._marks = self._get_marks()
+
+        def needs_eval(mark):
+            return mark.args or 'condition' in mark.kwargs
+
+        if self._marks:
+            self.result = False
+            # "holder" might be a MarkInfo or a MarkDecorator; only
+            # MarkInfo keeps track of all parameters it received in an
+            # _arglist attribute
+            for mark in self._marks:
+                self._mark = mark
+                if 'condition' in mark.kwargs:
+                    args = (mark.kwargs['condition'],)
+                else:
+                    args = mark.args
+
+                for expr in args:
+                    self.expr = expr
+                    if isinstance(expr, six.string_types):
+                        d = self._getglobals()
+                        result = cached_eval(self.item.config, expr, d)
+                    else:
+                        if "reason" not in mark.kwargs:
+                            # XXX better be checked at collection time
+                            msg = "you need to specify reason=STRING " \
+                                  "when using booleans as conditions."
+                            fail(msg)
+                        result = bool(expr)
+                    if result:
+                        self.result = True
+                        self.reason = mark.kwargs.get('reason', None)
                         self.expr = expr
-                        if isinstance(expr, six.string_types):
-                            d = self._getglobals()
-                            result = cached_eval(self.item.config, expr, d)
-                        else:
-                            if "reason" not in kwargs:
-                                # XXX better be checked at collection time
-                                msg = "you need to specify reason=STRING " \
-                                      "when using booleans as conditions."
-                                fail(msg)
-                            result = bool(expr)
-                        if result:
-                            self.result = True
-                            self.reason = kwargs.get('reason', None)
-                            self.expr = expr
-                            return self.result
-            else:
-                self.result = True
-        return getattr(self, 'result', False)
+                        return self.result
+
+                if not args:   
+                    self.result = True
+                    self.reason = mark.kwargs.get('reason', None)
+                    return self.result
+        return False
 
     def get(self, attr, default=None):
-        return self.holder.kwargs.get(attr, default)
+        if self._mark is None:
```

Comment:
Shouldn't this loop over all marks trying to get the `attr` (which is really a `kwarg` actually), returning `default` if no marks have it?

Can we rename it to `find_kwarg` instead for clarity btw? To actually understand what this does I had to add an `assert 0` and run the tests to figure out where it is called (because you can't actually do a search for "get").

Reply:
To exemplify:

```python
    def find_kwarg(self, kwarg, default=None):
        for mark in self._get_marks():
            if kwarg in mark.kwargs:
                return mark.kwargs[kwarg]
        return default
```

With this we can even get rid of `self._mark`.

Reply:
if one mark has run=True, and another one doesnt, we get wrong results, welcome to why marks are so horrific

Reply:
well, it started out as something much worse, i#ll try to whip it in shape more

Reply:
And about this suggestion @RonnyPfannschmidt?

Reply:
@nicoddemus that breaks current usage inside pytest at least

so in order to go there i'd prefer to keep that until we iterate it again